core/src/term/pattern/compile.rs

```diff
@@ -670,31 +671,48 @@ impl Compile for MatchData {
         if self.branches.iter().all(|(pat, _)| {
             matches!(
                 pat.data,
-                PatternData::Enum(EnumPattern { pattern: None, .. })
+                PatternData::Enum(EnumPattern { pattern: None, .. }) | PatternData::Wildcard
             )
         }) {
-            let tags_only = self.branches.into_iter().map(|(pat, body)| {
-                let PatternData::Enum(EnumPattern {tag, ..}) = pat.data else {
-                    panic!("match compilation: just tested that all cases are enum tags, but found a non enum tag pattern");
-                };
-
-                (tag, body)
-            }).collect();
+            // We take the first wildcard pattern as the default case. In theory, we could discard
+            // all the tags coming after the default branch; in practice we expect any sane match
+            // expression to have the default branch at the end, so this isn't a very useful
+            // optimization to care about.
```

Comment:
Unfortunately, I think you do need to discard the other tags. I just checked, and
```
'Bar |> match {
  'Foo => 1,
  _ => 2,
  'Bar => 3,
}
```
evaluates to 3, when it should evaluate to 2.

Reply:
Ah, right, this comment dates from the introduction of pattern matching and the specific optimization toward enums. It seems the accidental semantics has been that the catch-all case was independent of its position (and indeed, this was how it used to be parsed even). But now this is not the case anymore.

Good catch, thanks for noticing this!